exercises/05.actions/05.solution.history-revalidation/server/app.js

```diff
@@ -35,31 +55,32 @@ app.use((req, res, next) => {
 		next()
 	}
 })
+*/
 
 const moduleBasePath = new URL('../src', import.meta.url).href
 
-async function renderApp(res, returnValue) {
-	try {
-		const shipId = res.req.params.shipId || null
-		const search = res.req.query.search || ''
-		const data = { shipId, search }
-		shipDataStorage.run(data, () => {
-			const root = h(App)
-			const payload = { root, returnValue }
-			const { pipe } = renderToPipeableStream(payload, moduleBasePath)
-			pipe(res)
-		})
-	} catch (error) {
-		console.error(error)
-		res.status(500).json({ error: error.message })
-	}
+function renderApp(res, returnValue) {
+  const shipId = res.request.params.shipId || null
+  const search = res.request.query.search || ''
+  const data = { shipId, search }
+  return shipDataStorage.run(data, () => {
+    const root = h(App)
+    const payload = { root, returnValue }
+
+    const { pipe } = renderToPipeableStream(payload, moduleBasePath)
+    // TODO(mcollina): this is a workaround, it should not be needed but we are
+    // missing the primitive in the react-server-dom-esm package.
+    const stream = new PassThrough()
+    pipe(stream)
+    return stream
+  })
 }
 
-app.get('/rsc/:shipId?', async (req, res) => {
-	await renderApp(res, null)
+app.get('/rsc/:shipId?', (req, res) => {
+	res.type('text/html').send(renderApp(res, null))
 })
 
-app.post('/action/:shipId?', bodyParser.text(), async (req, res) => {
+app.post('/action/:shipId?', async (req, res) => {
```

Comment:
How do I trigger this action from the frontend? I couldn't find it.

Reply:
It's tricky. You click on the name of a ship and it changes to an edit form.